src/main/java/bio/terra/service/kubernetes/KubePodListener.java

```diff
@@ -0,0 +1,110 @@
+    private final Logger logger = LoggerFactory.getLogger(KubePodListener.class);
+    private final JobShutdownState jobShutdownState;
+    private final String namespace;
+    private final String thisPod;
+    private Exception exception;
+    private Map<String, Boolean> podMap; // pod name; true means running; false means deleted
+
+    public KubePodListener(JobShutdownState jobShutdownState, String namespace, String thisPod) {
+        this.jobShutdownState = jobShutdownState;
+        this.namespace = namespace;
+        this.thisPod = thisPod;
+        this.exception = null;
+
+        podMap = new HashMap<>();
+    }
+
+    @Override
+    public void run() {
+        logger.info("KubePodListener starting");
+
+        exception = null;
+        try {
+            ApiClient client = Config.defaultClient();
+            // infinite timeout
+            OkHttpClient httpClient =
+                client.getHttpClient().newBuilder().readTimeout(0, TimeUnit.SECONDS).build();
+            client.setHttpClient(httpClient);
+            Configuration.setDefaultApiClient(client);
+
+            CoreV1Api api = new CoreV1Api();
+
+            try (Watch<V1Namespace> watch =
+                     Watch.createWatch(
+                         client,
+                         api.listNamespacedPodCall(namespace,
+                             null,
+                             null,
+                             null,
+                             null,
+                             null,
+                             5,
+                             null,
+                             null,
+                             Boolean.TRUE,
+                             null),
+                         new TypeToken<Watch.Response<V1Namespace>>() { }.getType())) {
+                for (Watch.Response<V1Namespace> item : watch) {
+                    // If we are shutting down, we stop watching
+                    if (jobShutdownState.isShutdown()) {
+                        return;
+                    }
+
+                    String operation = item.type;
+                    String podName = item.object.getMetadata().getName();
+                    logger.info(String.format("%s : %s%n", operation, podName));
+
+                    if (StringUtils.contains(podName, API_POD_FILTER)) {
+                        if (StringUtils.equals(operation, "ADDED")) {
+                            logger.info("Added api pod: " + podName);
+                            podMap.put(podName, true);
+                        } else if (StringUtils.equals(operation, "DELETED")) {
+                            // TODO: call Stairway to inform it of a deleted pod
+                            Boolean deletedPod = podMap.get(podName);
+                            if (deletedPod != null && !deletedPod) {
+                                logger.info("Deleted api pod: " + podName);
+                                podMap.remove(podName);
```

Comment:
For deleted pods, should this instead update the hashmap to set the boolean value to false for this podName? 
Otherwise, could the hashmap instead just be an array of strings for running pods?

Reply:
Good catch! I had started with the list and then converted to a map. And missed this spot.